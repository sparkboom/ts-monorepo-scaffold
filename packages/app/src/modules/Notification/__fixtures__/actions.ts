import { showNotification, resetNotification } from '../actions';

export const showNotificationAction = showNotification('An error has occured', 'Scenario details here', 'ERROR', 1575163851945);

export const resetNotificationAction = resetNotification();
