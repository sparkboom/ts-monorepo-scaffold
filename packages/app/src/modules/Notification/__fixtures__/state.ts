import { NotificationState } from '../types';

export const nullState: NotificationState = {
  heading: null,
  details: null,
  level: null,
  expiryTime: null,
};

export const errorState: NotificationState = {
  heading: 'An error has occured',
  details: 'Scenario details here',
  level: 'ERROR',
  expiryTime: 1575163851945,
};

export const warnState: NotificationState = {
  heading: 'This is a warn heading',
  details: 'These are the warning details here',
  level: 'WARNING',
  expiryTime: 1575163851945,
};

export const successState: NotificationState = {
  heading: 'This is a success heading',
  details: 'These are the success details here',
  level: 'SUCCESS',
  expiryTime: 1575163851945,
};

export const infoState: NotificationState = {
  heading: 'This is an info heading',
  details: 'These are the info details here',
  level: 'INFO',
  expiryTime: 1575163851945,
};

export const noLevelState: NotificationState = {
  heading: 'This is a heading',
  details: 'These are the details here',
  level: null,
  expiryTime: 1575163851945,
};

export const infoNoHeadingState: NotificationState = {
  heading: null,
  details: 'These are the info details here',
  level: 'INFO',
  expiryTime: 1575163851945,
};
