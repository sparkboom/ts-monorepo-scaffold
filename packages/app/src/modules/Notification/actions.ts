/* eslint-disable import/prefer-default-export */
import { Action } from 'redux';
import { NotificationState } from './types';

// Constants
export const NOTIFICATION__NORMAL_EXPIRY_MS = 6000;
export const NOTIFICATION__LONG_EXPIRY_MS = 12000;
export const NOTIFICATION__XTRA_LONG_EXPIRY_MS = 500000;

// Types
interface ShowNotification {
  (heading: string, details?: string | null, level?: NotificationState['level'], lifespanMs?: number, date?: Date): AppAction<
    'NOTIFICATION__SHOW',
    NotificationState
  >;
}

// Actions
export const showNotification: ShowNotification = (heading, details = null, level = 'INFO', lifespanMs = NOTIFICATION__NORMAL_EXPIRY_MS, date) => ({
  type: 'NOTIFICATION__SHOW',
  payload: {
    heading,
    details,
    level,
    expiryTime: (date || new Date()).getTime() + lifespanMs,
  },
});

export const resetNotification = (): Action<'NOTIFICATION__RESET'> => ({
  type: 'NOTIFICATION__RESET',
});
