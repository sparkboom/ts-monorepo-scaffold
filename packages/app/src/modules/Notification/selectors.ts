import { NotificationState } from './types';

// Selectors
export const notificationSelector = (state: RootState): NotificationState => state.notification;
