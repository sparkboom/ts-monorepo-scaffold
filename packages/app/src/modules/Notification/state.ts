import { NotificationState } from './types';

const notificationState: NotificationState = {
  heading: null,
  details: null,
  level: null,
  expiryTime: null,
};

export default notificationState;
