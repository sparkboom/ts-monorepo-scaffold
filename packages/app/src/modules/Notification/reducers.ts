import { Reducer } from 'redux';
import * as actions from './actions';
import { NotificationState } from './types';
import configState from './state';

// Types
type NotificationActions = ActionUnion<typeof actions>;

// Reducer
const notificationReducer: Reducer<NotificationState, NotificationActions> = (state = configState, action) => {
  switch (action.type) {
    case 'NOTIFICATION__RESET': {
      return {
        heading: null,
        details: null,
        level: null,
        expiryTime: null,
      };
    }
    case 'NOTIFICATION__SHOW': {
      return {
        ...action.payload,
      };
    }
    default:
      return state;
  }
};

export default notificationReducer;
