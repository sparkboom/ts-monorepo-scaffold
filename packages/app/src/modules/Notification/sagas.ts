import { all, takeLatest, delay, race, take, call } from 'redux-saga/effects';
import { get } from 'lodash-es';
import { Action } from 'redux';
import { showNotification, resetNotification } from './actions';
import { putContext } from '../Logging/effects';
import { getTextSaga } from '../translations/sagas';
import { NotificationLevel } from './types';

// Constants
const LOG_CTX = 'NOTIFICATIONS.SAGAS';

// Types
type ShowNotification = ReturnType<typeof showNotification>;

// Sagas
export function* showNotificationSaga(action: ShowNotification) {
  const duration = Math.max(0, get(action, 'payload.expiryTime') - new Date().getTime() || 0);
  const [_, takeAction] = yield race([delay(duration), take('NOTIFICATION__RESET')]);

  if (!takeAction) {
    yield putContext(action, LOG_CTX)(resetNotification());
  }
}

/**
 * notifySaga
 *
 * Is a saga for convinience that can be used by other sagas. Passed text keys are translated into
 * localized text.
 *
 * @param action the current action being operated on by the sagas
 * @param headingTextKey the text key for the heading
 * @param detailsTextKey the text key for the details, optional
 * @param level the notification level, this describes the style of the notification.
 */
export function* notifySaga(
  action: Action<string>,
  headingTextKey: string,
  detailsTextKey: string | null,
  level: NotificationLevel = 'INFO',
  lifespanMs?: number,
): any {
  const put = putContext(action, LOG_CTX);
  let submittedHeading = '';
  let details = '';
  try {
    const headingTextEntry = { key: headingTextKey };
    submittedHeading = yield call(getTextSaga, [headingTextEntry], action);
    if (detailsTextKey) {
      const detailsTextEntry = { key: detailsTextKey };
      details = yield call(getTextSaga, [detailsTextEntry], action);
    }
    yield put(showNotification(submittedHeading, details, level, lifespanMs));
  } catch (err) {
    yield put.errorWarn(err);
  }
}

function* notificationSaga() {
  yield all([takeLatest('NOTIFICATION__SHOW', showNotificationSaga)]);
}

export default notificationSaga;
