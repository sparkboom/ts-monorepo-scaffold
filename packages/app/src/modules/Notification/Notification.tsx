import React, { FunctionComponent, useCallback } from 'react';
import { isString } from 'lodash-es';
import Toast from '@salesforce/design-system-react/components/toast';
import ToastContainer from '@salesforce/design-system-react/components/toast/container';
import { notificationSelector } from './selectors';
import { useSelector, useDispatch } from 'react-redux';
import { useIntl } from 'react-intl';
import { resetNotification } from './actions';

// Types
interface NotificationProps {
  className?: string;
  children?: never;
}

// Components
const Notification: FunctionComponent<NotificationProps> = (props) => {
  const { className } = props;
  const notification = useSelector(notificationSelector);
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();


  const assistiveText = {
    closeButton: formatMessage({ id: 'common.notification.closeButton.assistive' }),
  };
  const labels = {
    heading: notification.heading,
    details: notification.details || undefined,
  };
  const toastLevel = isString(notification.level) ? notification.level.toLowerCase() : 'info';

  const handleRequestClose = useCallback(() => dispatch(resetNotification()), [dispatch]);

  if (!labels.heading) {
    return null;
  }

  return (
    <ToastContainer className={className}>
      <Toast assistiveText={assistiveText} labels={labels} variant={toastLevel} onRequestClose={handleRequestClose} />
    </ToastContainer>
  );
};
Notification.displayName = 'ASLCNotification';

export default Notification;
