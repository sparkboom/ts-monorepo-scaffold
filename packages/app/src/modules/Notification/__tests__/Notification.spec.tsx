import { shallow } from 'enzyme';
import { mocked } from 'ts-jest/utils';
import React from 'react';
import Notification from '../Notification';
import { useSelector, useDispatch } from 'react-redux';
import { infoState, errorState, warnState, successState, noLevelState, infoNoHeadingState } from '../__fixtures__/state';
import { resetNotificationAction } from '../__fixtures__/actions';

// Mocks
const dispatch = useDispatch();
const mockedUseSelector = mocked(useSelector, false);
const mockedDispatch = mocked(dispatch, false);

// Tests
describe('Notification > Notification', () => {
  describe('App Component', () => {
    afterEach(() => {
      mockedUseSelector.mockReset();
    });

    it('should render the heading', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('labels').heading).toEqual(infoState.heading);
    });

    it('should render the details', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('labels').details).toEqual(infoState.details);
    });

    it('should correctly render a notification with an error level', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(errorState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('variant')).toEqual('error');
    });

    it('should correctly render a notification with a warn level', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(warnState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('variant')).toEqual('warning');
    });

    it('should correctly render a notification with a success level', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(successState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('variant')).toEqual('success');
    });

    it('should correctly render a notification with an info level', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('variant')).toEqual('info');
    });

    it('should render an info notification if no level is provided', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(noLevelState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('variant')).toEqual('info');
    });

    it('should not render if a heading is not provided', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoNoHeadingState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.isEmptyRender()).toBeTruthy();
    });

    it('should correctly apply assistive texts', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper.find('Toast').prop<any>('assistiveText')).toEqual({
        closeButton: '{text @ common.notification.closeButton.assistive}',
      });
    });

    it('should dispatch a reset notification action if request close is triggered', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(infoState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);
      wrapper.find('Toast').prop<any>('onRequestClose')();

      // Assert
      expect(mockedDispatch.mock.calls[0][0]).toEqual(resetNotificationAction);
    });

    it('should match snapshot when rendering a success notification', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(successState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when rendering an error notification', () => {
      // Arrange
      mockedUseSelector.mockReturnValue(errorState);

      // Act
      const wrapper = shallow(<Notification className="my-class" />);

      // Assert
      expect(wrapper).toMatchSnapshot();
    });
  });
});
