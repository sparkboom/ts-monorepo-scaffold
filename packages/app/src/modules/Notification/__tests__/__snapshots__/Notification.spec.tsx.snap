// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification > Notification App Component should match snapshot when rendering a success notification 1`] = `
<ToastContainer
  className="my-class"
>
  <Toast
    assistiveText={
      Object {
        "closeButton": "{text @ common.notification.closeButton.assistive}",
      }
    }
    labels={
      Object {
        "details": "These are the success details here",
        "heading": "This is a success heading",
      }
    }
    onRequestClose={[Function]}
    variant="success"
  />
</ToastContainer>
`;

exports[`Notification > Notification App Component should match snapshot when rendering an error notification 1`] = `
<ToastContainer
  className="my-class"
>
  <Toast
    assistiveText={
      Object {
        "closeButton": "{text @ common.notification.closeButton.assistive}",
      }
    }
    labels={
      Object {
        "details": "Scenario details here",
        "heading": "An error has occured",
      }
    }
    onRequestClose={[Function]}
    variant="error"
  />
</ToastContainer>
`;
